### DISCUSSION TITLE
How to identify when the next.js hydration finishes hydration on client side?

### DISCUSSION THREAD

**Discussion by bhavin7181** [vercel/next.js#18271] [Help] [answered]
We need to change DOM content on certain html elements when the next.js finishes hydration of the page. Is there an event which we can listen to make this change post-hydration? 
Thank you in advance.

**Comment by jamesmosier** [comment, marked as answer]
You should be able to use `useEffect` for this! `useEffect` does not run on the server, so that would be a good choice to run some code after the client loads.

**Reply by danielGz** [reply]
`useEffect `does run before hydration finishes though, there are cases in which you need to check for additional stuff within useEffect, such is the case of `router.query`. The docs mention this: _"The query string parsed to an object. It will be an empty object during prerendering if the page doesn't have data fetching requirements."_

And I ran into the case of having an empty dependencies array and trying to do something with `router.query` but It'd never execute again by the time query was populated.

**Reply by IzaacBarratt** [reply]
Also have this exact problem with query parameters. Any suggestions?

**Reply by mdupree** [reply]
why not use `router.isReady`

**Reply by alexeysarakuz** [reply]
you saved my day @mdupree! Thanks!

**Comment by atellmer** [comment]
useEffect runs even when hydration is not yet complete. At least in next.js 14 and react 18. 

I don't know who came up with this brilliant (not) idea.
